diff --git a/dist/ios/spm.js b/dist/ios/spm.js
index 1c1e4af26c2475d803f5c5eaaa957b333d0453df..bba232f88674ee831bcf678a22b24c4a727107c6 100644
--- a/dist/ios/spm.js
+++ b/dist/ios/spm.js
@@ -138,14 +138,14 @@ function addSPMPackageToProject(project, targetId, pkg, projectRoot) {
         comment: packageReferenceComment,
     });
     for (const lib of pkg.libs) {
-        const { uuid: spmProductDependencyUUID } = helper.addOrUpdateEntry('XCSwiftPackageProductDependency', lib, {
+        const { uuid: spmProductDependencyUUID } = helper.addEntry('XCSwiftPackageProductDependency', lib, {
             isa: 'XCSwiftPackageProductDependency',
             package: spmPackageReferenceUUID,
             package_comment: spmPackageReferenceComment,
             productName: lib,
         });
         const libComment = `${lib} in Frameworks`;
-        const { uuid: spmBuildFileUuid } = helper.addOrUpdateEntry('PBXBuildFile', libComment, {
+        const { uuid: spmBuildFileUuid } = helper.addEntry('PBXBuildFile', libComment, {
             isa: 'PBXBuildFile',
             productRef: spmProductDependencyUUID,
             productRef_comment: lib,
@@ -184,6 +184,17 @@ class SPMHelper {
             comment: entryComment,
         };
     }
+    addEntry(section, entryComment, entry) {
+        const pbxSection = this.getOrCreateSection(section);
+        const entryUuid = this.pbxProject.generateUuid();
+        const entryCommentKey = `${entryUuid}_comment`;
+        pbxSection[entryCommentKey] = entryComment;
+        pbxSection[entryUuid] = entry;
+        return {
+            uuid: entryUuid,
+            comment: entryComment,
+        };
+    }
     getExistingOrGenerateUUID(section, comment) {
         var _a;
         const existingUUID = (_a = Object.keys(this.pbxProject.hash.project.objects[section])
