diff --git a/lib/pbxFile.js b/lib/pbxFile.js
index 1b52d259bc4b5b3dcf935cb3f8c2f789bb5ddbc7..1c140853dd05dc94681b53bc02c7aa54a702603a 100644
--- a/lib/pbxFile.js
+++ b/lib/pbxFile.js
@@ -128,9 +128,9 @@ function pbxFile(filepath, opt) {
     if (opt.explicitFileType) {
         this.explicitFileType = opt.explicitFileType;
         this.basename = this.basename + '.' + defaultExtension(this);
-        delete this.path;
+        // delete this.path;
         delete this.lastKnownFileType;
-        delete this.group;
+        // delete this.group;
         delete this.defaultEncoding;
     }
 
diff --git a/lib/pbxProject.js b/lib/pbxProject.js
index 70f82ea1a1c41de3ea18b66e4d89584c16a79e88..aa0096f47d9111c7d6e0abdaf5749348cd3da8f1 100644
--- a/lib/pbxProject.js
+++ b/lib/pbxProject.js
@@ -119,8 +119,8 @@ pbxProject.prototype.addPluginFile = function(path, opt) {
     file.plugin = true; // durr
     correctForPluginsPath(file, this);
 
-    // null is better for early errors
-    if (this.hasFile(file.path)) return null;
+    const existingFile = this.hasFile(file.path);
+    if (existingFile) return file;
 
     file.fileRef = this.generateUuid();
 
@@ -184,8 +184,9 @@ pbxProject.prototype.addSourceFile = function (path, opt, group) {
 
     file.target = opt ? opt.target : undefined;
     file.uuid = this.generateUuid();
-
-    this.addToPbxBuildFileSection(file);        // PBXBuildFile
+    if(file.fileRef) {
+        this.addToPbxBuildFileSection(file);        // PBXBuildFile
+    }
     this.addToPbxSourcesBuildPhase(file);       // PBXSourcesBuildPhase
 
     return file;
@@ -256,19 +257,23 @@ pbxProject.prototype.addResourceFile = function(path, opt, group) {
     opt = opt || {};
 
     var file;
-
+    var fileIsAlreadyAdded = false;
     if (opt.plugin) {
         file = this.addPluginFile(path, opt);
         if (!file) return false;
     } else {
         file = new pbxFile(path, opt);
-        if (this.hasFile(file.path)) return false;
+        const existingFile = this.hasFile(file.path);
+        fileIsAlreadyAdded = !!existingFile;
+        if (existingFile) {
+            file.fileRef = existingFile.fileRef;
+        }
     }
 
     file.uuid = this.generateUuid();
     file.target = opt ? opt.target : undefined;
 
-    if (!opt.plugin) {
+    if (!opt.plugin && !fileIsAlreadyAdded) {
         correctForResourcesPath(file, this);
         file.fileRef = this.generateUuid();
     }
@@ -278,7 +283,7 @@ pbxProject.prototype.addResourceFile = function(path, opt, group) {
         this.addToPbxResourcesBuildPhase(file);     // PBXResourcesBuildPhase
     }
 
-    if (!opt.plugin) {
+    if (!opt.plugin && !fileIsAlreadyAdded) {
         this.addToPbxFileReferenceSection(file);    // PBXFileReference
         if (group) {
             if (this.getPBXGroupByKey(group)) {
@@ -859,20 +864,20 @@ pbxProject.prototype.removeFromPluginsPbxGroup = function(file) {
     }
 }
 
-pbxProject.prototype.addToResourcesPbxGroup = function(file) {
-    var pluginsGroup = this.pbxGroupByName('Resources');
+pbxProject.prototype.addToResourcesPbxGroup = function(file, groupName = 'Resources') {
+    var pluginsGroup = this.pbxGroupByName(groupName);
     if (!pluginsGroup) {
-        this.addPbxGroup([file.path], 'Resources');
+        this.addPbxGroup([file.path], groupName);
     } else {
         pluginsGroup.children.push(pbxGroupChild(file));
     }
 }
 
-pbxProject.prototype.removeFromResourcesPbxGroup = function(file) {
-    if (!this.pbxGroupByName('Resources')) {
+pbxProject.prototype.removeFromResourcesPbxGroup = function(file, groupName = 'Resources') {
+    if (!this.pbxGroupByName(groupName)) {
         return null;
     }
-    var pluginsGroupChildren = this.pbxGroupByName('Resources').children, i;
+    var pluginsGroupChildren = this.pbxGroupByName(groupName).children, i;
     for (i in pluginsGroupChildren) {
         if (pbxGroupChild(file).value == pluginsGroupChildren[i].value &&
             pbxGroupChild(file).comment == pluginsGroupChildren[i].comment) {
@@ -916,6 +921,7 @@ function getReferenceInPbxBuildFile(buildFileReferences, fileReference) {
 }
 
 pbxProject.prototype.addToPbxEmbedFrameworksBuildPhase = function (file) {
+    console.error('addToPbxEmbedFrameworksBuildPhase', JSON.stringify(file));
     var sources = this.pbxEmbedFrameworksBuildPhaseObj(file.target);
 
     if (sources) {
@@ -1363,7 +1369,6 @@ pbxProject.prototype.buildPhaseObject = function(name, group, target) {
     var section = this.hash.project.objects[name],
         obj, sectionKey, key;
     var buildPhase = this.buildPhase(group, target);
-
     for (key in section) {
 
         // only look for comments
@@ -1680,7 +1685,7 @@ pbxProject.prototype.hasFile = function(filePath) {
     for (id in files) {
         file = files[id];
         if (file.path == filePath || file.path == ('"' + filePath + '"')) {
-            return file;
+            return {...file,  fileRef: id};
         }
     }
 
@@ -1700,11 +1705,21 @@ pbxProject.prototype.getFileKey = function(filePath) {
     return false;
 }
 
-pbxProject.prototype.addTarget = function(name, type, subfolder, parentTarget) {
+pbxProject.prototype.getProductFile = function(target) {
+    var productReferenceUuid = target.pbxNativeTarget.productReference;
+    const fileReferenceSection = this.pbxFileReferenceSection();
+    const productReference = fileReferenceSection[productReferenceUuid];
+    const productFile = new pbxFile(productReference.path);
+    productFile.fileRef = productReferenceUuid;
+    productFile.uuid = productReferenceUuid;
+    return productFile;
+}
+pbxProject.prototype.addTarget = function(name, type, subfolder, parentTarget, productTargetType) {
 
     // Setup uuid and name of new target
     var targetUuid = this.generateUuid(),
         targetType = type,
+        productTargetType = productTargetType || targetType,
         targetSubfolder = subfolder || name,
         targetName = name.trim();
 
@@ -1753,7 +1768,7 @@ pbxProject.prototype.addTarget = function(name, type, subfolder, parentTarget) {
 
     // Product: Create
     var productName = targetName,
-        productType = producttypeForTargettype(targetType),
+        productType = producttypeForTargettype(productTargetType),
         productFileType = filetypeForProducttype(productType),
         productFile = this.addProductFile(productName, { group: 'Copy Files', 'target': targetUuid, 'explicitFileType': productFileType}),
         productFileName = productFile.basename;
@@ -1781,10 +1796,10 @@ pbxProject.prototype.addTarget = function(name, type, subfolder, parentTarget) {
     // Target: Add to PBXNativeTarget section
     this.addToPbxNativeTargetSection(target);
 
-    if (targetType === 'app_extension' || targetType === 'watch_extension' || targetType === 'watch_app') {
-        const isWatchApp = targetType === 'watch_app';
+    if (productTargetType === 'app_extension' || productTargetType === 'watch_extension' || productTargetType === 'watch_app') {
+        const isWatchApp = productTargetType === 'watch_app';
         const copyTargetUuid = parentTarget || this.getFirstTarget().uuid;
-        const phaseComment = targetType === 'watch_app' ? 'Embed Watch Content' : 'Copy Files';
+        const phaseComment = productTargetType === 'watch_app' ? 'Embed Watch Content' : 'Copy Files';
         let destination;
 
         if(isWatchApp) {
@@ -1792,7 +1807,7 @@ pbxProject.prototype.addTarget = function(name, type, subfolder, parentTarget) {
         }
 
         // Create CopyFiles phase in parent target
-        this.addBuildPhase([], 'PBXCopyFilesBuildPhase', phaseComment, copyTargetUuid,  targetType, destination);
+        this.addBuildPhase([], 'PBXCopyFilesBuildPhase', phaseComment, copyTargetUuid,  productTargetType, destination);
 
         // Add product to CopyFiles phase
         this.addToPbxCopyfilesBuildPhase(productFile, phaseComment, copyTargetUuid);
@@ -2058,7 +2073,7 @@ function pbxFileReferenceObj(file) {
         fileObject.name = "\"" + fileObject.name + "\"";
     }
 
-    if(!file.path.match(NO_SPECIAL_SYMBOLS)) {
+    if(file.path && !file.path.match(NO_SPECIAL_SYMBOLS)) {
         fileObject.path = "\"" + fileObject.path + "\"";
     }
 
@@ -2525,10 +2540,11 @@ pbxProject.prototype.getPBXObject = function(name) {
 pbxProject.prototype.addFile = function (path, group, opt) {
     var file = new pbxFile(path, opt);
 
-    // null is better for early errors
-    if (this.hasFile(file.path)) return null;
+    const existingFile = this.hasFile(file.path);
+    if (existingFile) return {...file,  fileRef: existingFile.fileRef};
 
     file.fileRef = this.generateUuid();
+    file.target = opt ? opt.target : undefined;
 
     this.addToPbxFileReferenceSection(file);    // PBXFileReference
 
